// @ai-sdk/replicate/dist/index.mjs
// src/replicate-provider.ts
import {
  NoSuchModelError
} from "@ai-sdk/provider";
import {
  loadApiKey,
  withUserAgentSuffix
} from "@ai-sdk/provider-utils";

// src/replicate-image-model.ts
import {
  combineHeaders,
  convertImageModelFileToDataUri,
  createBinaryResponseHandler,
  createJsonResponseHandler,
  getFromApi,
  lazySchema,
  parseProviderOptions,
  postJsonToApi,
  resolve,
  zodSchema
} from "@ai-sdk/provider-utils";
import { z as z2 } from "zod/v4";

// src/replicate-error.ts
import { createJsonErrorResponseHandler } from "@ai-sdk/provider-utils";
import { z } from "zod/v4";
var replicateErrorSchema = z.object({
  detail: z.string().optional(),
  error: z.string().optional()
});
var replicateFailedResponseHandler = createJsonErrorResponseHandler({
  errorSchema: replicateErrorSchema,
  errorToMessage: (error) => {
    var _a, _b;
    return (_b = (_a = error.detail) != null ? _a : error.error) != null ? _b : "Unknown Replicate error";
  }
});

// src/replicate-image-model.ts
var FLUX_2_MODEL_PATTERN = /^black-forest-labs\/flux-2-/;
var MAX_FLUX_2_INPUT_IMAGES = 8;
var ReplicateImageModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v3";
  }
  get maxImagesPerCall() {
    return this.isFlux2Model ? MAX_FLUX_2_INPUT_IMAGES : 1;
  }
  get provider() {
    return this.config.provider;
  }
  get isFlux2Model() {
    return FLUX_2_MODEL_PATTERN.test(this.modelId);
  }
  async doGenerate({
    prompt,
    n,
    aspectRatio,
    size,
    seed,
    providerOptions,
    headers,
    abortSignal,
    files,
    mask
  }) {
    var _a, _b, _c;
    const warnings = [];
    const [modelId, version] = this.modelId.split(":");
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const replicateOptions = await parseProviderOptions({
      provider: "replicate",
      providerOptions,
      schema: replicateImageModelOptionsSchema
    });
    let imageInputs = {};
    if (files != null && files.length > 0) {
      if (this.isFlux2Model) {
        for (let i = 0; i < Math.min(files.length, MAX_FLUX_2_INPUT_IMAGES); i++) {
          const key = i === 0 ? "input_image" : `input_image_${i + 1}`;
          imageInputs[key] = convertImageModelFileToDataUri(files[i]);
        }
        if (files.length > MAX_FLUX_2_INPUT_IMAGES) {
          warnings.push({
            type: "other",
            message: `Flux-2 models support up to ${MAX_FLUX_2_INPUT_IMAGES} input images. Additional images are ignored.`
          });
        }
      } else {
        imageInputs = { image: convertImageModelFileToDataUri(files[0]) };
        if (files.length > 1) {
          warnings.push({
            type: "other",
            message: "This Replicate model only supports a single input image. Additional images are ignored."
          });
        }
      }
    }
    let maskInput;
    if (mask != null) {
      if (this.isFlux2Model) {
        warnings.push({
          type: "other",
          message: "Flux-2 models do not support mask input. The mask will be ignored."
        });
      } else {
        maskInput = convertImageModelFileToDataUri(mask);
      }
    }
    const { maxWaitTimeInSeconds, ...inputOptions } = replicateOptions != null ? replicateOptions : {};
    const preferHeader = maxWaitTimeInSeconds != null ? { prefer: `wait=${maxWaitTimeInSeconds}` } : { prefer: "wait" };
    const {
      value: { output },
      responseHeaders
    } = await postJsonToApi({
      url: (
        // different endpoints for versioned vs unversioned models:
        version != null ? `${this.config.baseURL}/predictions` : `${this.config.baseURL}/models/${modelId}/predictions`
      ),
      headers: combineHeaders(
        await resolve(this.config.headers),
        headers,
        preferHeader
      ),
      body: {
        input: {
          prompt,
          aspect_ratio: aspectRatio,
          size,
          seed,
          num_outputs: n,
          ...imageInputs,
          ...maskInput != null ? { mask: maskInput } : {},
          ...inputOptions
        },
        // for versioned models, include the version in the body:
        ...version != null ? { version } : {}
      },
      successfulResponseHandler: createJsonResponseHandler(
        replicateImageResponseSchema
      ),
      failedResponseHandler: replicateFailedResponseHandler,
      abortSignal,
      fetch: this.config.fetch
    });
    const outputArray = Array.isArray(output) ? output : [output];
    const images = await Promise.all(
      outputArray.map(async (url) => {
        const { value: image } = await getFromApi({
          url,
          successfulResponseHandler: createBinaryResponseHandler(),
          failedResponseHandler: replicateFailedResponseHandler,
          abortSignal,
          fetch: this.config.fetch
        });
        return image;
      })
    );
    return {
      images,
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders
      }
    };
  }
};
var replicateImageResponseSchema = z2.object({
  output: z2.union([z2.array(z2.string()), z2.string()])
});
var replicateImageModelOptionsSchema = lazySchema(
  () => zodSchema(
    z2.object({
      /**
       * Maximum time in seconds to wait for the prediction to complete in sync mode.
       * By default, Replicate uses sync mode with a 60-second timeout.
       *
       * - When not specified: Uses default 60-second sync wait (`prefer: wait`)
       * - When set to a positive number: Uses that duration (`prefer: wait=N`)
       */
      maxWaitTimeInSeconds: z2.number().positive().nullish(),
      /**
       * Guidance scale for classifier-free guidance.
       * Higher values make the output more closely match the prompt.
       */
      guidance_scale: z2.number().nullish(),
      /**
       * Number of denoising steps. More steps = higher quality but slower.
       */
      num_inference_steps: z2.number().nullish(),
      /**
       * Negative prompt to guide what to avoid in the generation.
       */
      negative_prompt: z2.string().nullish(),
      /**
       * Output image format.
       */
      output_format: z2.enum(["png", "jpg", "webp"]).nullish(),
      /**
       * Output image quality (1-100). Only applies to jpg and webp.
       */
      output_quality: z2.number().min(1).max(100).nullish(),
      /**
       * Strength of the transformation for img2img (0-1).
       * Lower values keep more of the original image.
       */
      strength: z2.number().min(0).max(1).nullish()
    }).passthrough()
  )
);

// src/replicate-video-model.ts
import {
  AISDKError
} from "@ai-sdk/provider";
import {
  combineHeaders as combineHeaders2,
  convertImageModelFileToDataUri as convertImageModelFileToDataUri2,
  createJsonResponseHandler as createJsonResponseHandler2,
  delay,
  getFromApi as getFromApi2,
  lazySchema as lazySchema2,
  parseProviderOptions as parseProviderOptions2,
  postJsonToApi as postJsonToApi2,
  resolve as resolve2,
  zodSchema as zodSchema2
} from "@ai-sdk/provider-utils";
import { z as z3 } from "zod/v4";
var ReplicateVideoModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v3";
    this.maxVideosPerCall = 1;
  }
  // Replicate video models support 1 video at a time
  get provider() {
    return this.config.provider;
  }
  async doGenerate(options) {
    var _a, _b, _c, _d, _e, _f, _g;
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const warnings = [];
    const replicateOptions = await parseProviderOptions2({
      provider: "replicate",
      providerOptions: options.providerOptions,
      schema: replicateVideoModelOptionsSchema
    });
    const [modelId, version] = this.modelId.split(":");
    const input = {};
    if (options.prompt != null) {
      input.prompt = options.prompt;
    }
    if (options.image != null) {
      if (options.image.type === "url") {
        input.image = options.image.url;
      } else {
        input.image = convertImageModelFileToDataUri2(options.image);
      }
    }
    if (options.aspectRatio) {
      input.aspect_ratio = options.aspectRatio;
    }
    if (options.resolution) {
      input.size = options.resolution;
    }
    if (options.duration) {
      input.duration = options.duration;
    }
    if (options.fps) {
      input.fps = options.fps;
    }
    if (options.seed) {
      input.seed = options.seed;
    }
    if (replicateOptions != null) {
      const opts = replicateOptions;
      if (opts.guidance_scale !== void 0 && opts.guidance_scale !== null) {
        input.guidance_scale = opts.guidance_scale;
      }
      if (opts.num_inference_steps !== void 0 && opts.num_inference_steps !== null) {
        input.num_inference_steps = opts.num_inference_steps;
      }
      if (opts.motion_bucket_id !== void 0 && opts.motion_bucket_id !== null) {
        input.motion_bucket_id = opts.motion_bucket_id;
      }
      if (opts.cond_aug !== void 0 && opts.cond_aug !== null) {
        input.cond_aug = opts.cond_aug;
      }
      if (opts.decoding_t !== void 0 && opts.decoding_t !== null) {
        input.decoding_t = opts.decoding_t;
      }
      if (opts.video_length !== void 0 && opts.video_length !== null) {
        input.video_length = opts.video_length;
      }
      if (opts.sizing_strategy !== void 0 && opts.sizing_strategy !== null) {
        input.sizing_strategy = opts.sizing_strategy;
      }
      if (opts.frames_per_second !== void 0 && opts.frames_per_second !== null) {
        input.frames_per_second = opts.frames_per_second;
      }
      if (opts.prompt_optimizer !== void 0 && opts.prompt_optimizer !== null) {
        input.prompt_optimizer = opts.prompt_optimizer;
      }
      for (const [key, value] of Object.entries(opts)) {
        if (![
          "pollIntervalMs",
          "pollTimeoutMs",
          "maxWaitTimeInSeconds",
          "guidance_scale",
          "num_inference_steps",
          "motion_bucket_id",
          "cond_aug",
          "decoding_t",
          "video_length",
          "sizing_strategy",
          "frames_per_second",
          "prompt_optimizer"
        ].includes(key)) {
          input[key] = value;
        }
      }
    }
    const maxWaitTimeInSeconds = replicateOptions == null ? void 0 : replicateOptions.maxWaitTimeInSeconds;
    const preferHeader = maxWaitTimeInSeconds != null ? { prefer: `wait=${maxWaitTimeInSeconds}` } : { prefer: "wait" };
    const predictionUrl = version != null ? `${this.config.baseURL}/predictions` : `${this.config.baseURL}/models/${modelId}/predictions`;
    const { value: prediction, responseHeaders } = await postJsonToApi2({
      url: predictionUrl,
      headers: combineHeaders2(
        await resolve2(this.config.headers),
        options.headers,
        preferHeader
      ),
      body: {
        input,
        ...version != null ? { version } : {}
      },
      successfulResponseHandler: createJsonResponseHandler2(
        replicatePredictionSchema
      ),
      failedResponseHandler: replicateFailedResponseHandler,
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    let finalPrediction = prediction;
    if (prediction.status === "starting" || prediction.status === "processing") {
      const pollIntervalMs = (_d = replicateOptions == null ? void 0 : replicateOptions.pollIntervalMs) != null ? _d : 2e3;
      const pollTimeoutMs = (_e = replicateOptions == null ? void 0 : replicateOptions.pollTimeoutMs) != null ? _e : 3e5;
      const startTime = Date.now();
      while (finalPrediction.status === "starting" || finalPrediction.status === "processing") {
        if (Date.now() - startTime > pollTimeoutMs) {
          throw new AISDKError({
            name: "REPLICATE_VIDEO_GENERATION_TIMEOUT",
            message: `Video generation timed out after ${pollTimeoutMs}ms`
          });
        }
        await delay(pollIntervalMs);
        if ((_f = options.abortSignal) == null ? void 0 : _f.aborted) {
          throw new AISDKError({
            name: "REPLICATE_VIDEO_GENERATION_ABORTED",
            message: "Video generation request was aborted"
          });
        }
        const { value: statusPrediction } = await getFromApi2({
          url: finalPrediction.urls.get,
          headers: await resolve2(this.config.headers),
          successfulResponseHandler: createJsonResponseHandler2(
            replicatePredictionSchema
          ),
          failedResponseHandler: replicateFailedResponseHandler,
          abortSignal: options.abortSignal,
          fetch: this.config.fetch
        });
        finalPrediction = statusPrediction;
      }
    }
    if (finalPrediction.status === "failed") {
      throw new AISDKError({
        name: "REPLICATE_VIDEO_GENERATION_FAILED",
        message: `Video generation failed: ${(_g = finalPrediction.error) != null ? _g : "Unknown error"}`
      });
    }
    if (finalPrediction.status === "canceled") {
      throw new AISDKError({
        name: "REPLICATE_VIDEO_GENERATION_CANCELED",
        message: "Video generation was canceled"
      });
    }
    const videoUrl = finalPrediction.output;
    if (!videoUrl) {
      throw new AISDKError({
        name: "REPLICATE_VIDEO_GENERATION_ERROR",
        message: "No video URL in response"
      });
    }
    return {
      videos: [
        {
          type: "url",
          url: videoUrl,
          mediaType: "video/mp4"
        }
      ],
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders
      },
      providerMetadata: {
        replicate: {
          videos: [
            {
              url: videoUrl
            }
          ],
          predictionId: finalPrediction.id,
          metrics: finalPrediction.metrics
        }
      }
    };
  }
};
var replicatePredictionSchema = z3.object({
  id: z3.string(),
  status: z3.enum(["starting", "processing", "succeeded", "failed", "canceled"]),
  output: z3.string().nullish(),
  error: z3.string().nullish(),
  urls: z3.object({
    get: z3.string()
  }),
  metrics: z3.object({
    predict_time: z3.number().nullish()
  }).nullish()
});
var replicateVideoModelOptionsSchema = lazySchema2(
  () => zodSchema2(
    z3.object({
      pollIntervalMs: z3.number().positive().nullish(),
      pollTimeoutMs: z3.number().positive().nullish(),
      maxWaitTimeInSeconds: z3.number().positive().nullish(),
      guidance_scale: z3.number().nullish(),
      num_inference_steps: z3.number().nullish(),
      motion_bucket_id: z3.number().nullish(),
      cond_aug: z3.number().nullish(),
      decoding_t: z3.number().nullish(),
      video_length: z3.string().nullish(),
      sizing_strategy: z3.string().nullish(),
      frames_per_second: z3.number().nullish(),
      prompt_optimizer: z3.boolean().nullish()
    }).loose()
  )
);

// src/version.ts
var VERSION = true ? "2.0.33" : "0.0.0-test";

// src/replicate-provider.ts
function createReplicate(options = {}) {
  const getHeaders = () => withUserAgentSuffix(
    {
      Authorization: `Bearer ${loadApiKey({
        apiKey: options.apiToken,
        environmentVariableName: "REPLICATE_API_TOKEN",
        description: "Replicate"
      })}`,
      ...options.headers
    },
    `ai-sdk/replicate/${VERSION}`
  );
  const createImageModel = (modelId) => {
    var _a;
    return new ReplicateImageModel(modelId, {
      provider: "replicate",
      baseURL: (_a = options.baseURL) != null ? _a : "https://api.replicate.com/v1",
      headers: getHeaders(),
      fetch: options.fetch
    });
  };
  const createVideoModel = (modelId) => {
    var _a;
    return new ReplicateVideoModel(modelId, {
      provider: "replicate.video",
      baseURL: (_a = options.baseURL) != null ? _a : "https://api.replicate.com/v1",
      headers: getHeaders,
      fetch: options.fetch
    });
  };
  const embeddingModel = (modelId) => {
    throw new NoSuchModelError({
      modelId,
      modelType: "embeddingModel"
    });
  };
  return {
    specificationVersion: "v3",
    image: createImageModel,
    imageModel: createImageModel,
    languageModel: (modelId) => {
      throw new NoSuchModelError({
        modelId,
        modelType: "languageModel"
      });
    },
    embeddingModel,
    textEmbeddingModel: embeddingModel,
    video: createVideoModel,
    videoModel: createVideoModel
  };
}
var replicate = createReplicate();
export {
  VERSION,
  createReplicate,
  replicate
};
