// @ai-sdk/replicate/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  VERSION: () => VERSION,
  createReplicate: () => createReplicate,
  replicate: () => replicate
});
module.exports = __toCommonJS(index_exports);

// src/replicate-provider.ts
var import_provider2 = require("@ai-sdk/provider");
var import_provider_utils4 = require("@ai-sdk/provider-utils");

// src/replicate-image-model.ts
var import_provider_utils2 = require("@ai-sdk/provider-utils");
var import_v42 = require("zod/v4");

// src/replicate-error.ts
var import_provider_utils = require("@ai-sdk/provider-utils");
var import_v4 = require("zod/v4");
var replicateErrorSchema = import_v4.z.object({
  detail: import_v4.z.string().optional(),
  error: import_v4.z.string().optional()
});
var replicateFailedResponseHandler = (0, import_provider_utils.createJsonErrorResponseHandler)({
  errorSchema: replicateErrorSchema,
  errorToMessage: (error) => {
    var _a, _b;
    return (_b = (_a = error.detail) != null ? _a : error.error) != null ? _b : "Unknown Replicate error";
  }
});

// src/replicate-image-model.ts
var FLUX_2_MODEL_PATTERN = /^black-forest-labs\/flux-2-/;
var MAX_FLUX_2_INPUT_IMAGES = 8;
var ReplicateImageModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v3";
  }
  get maxImagesPerCall() {
    return this.isFlux2Model ? MAX_FLUX_2_INPUT_IMAGES : 1;
  }
  get provider() {
    return this.config.provider;
  }
  get isFlux2Model() {
    return FLUX_2_MODEL_PATTERN.test(this.modelId);
  }
  async doGenerate({
    prompt,
    n,
    aspectRatio,
    size,
    seed,
    providerOptions,
    headers,
    abortSignal,
    files,
    mask
  }) {
    var _a, _b, _c;
    const warnings = [];
    const [modelId, version] = this.modelId.split(":");
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const replicateOptions = await (0, import_provider_utils2.parseProviderOptions)({
      provider: "replicate",
      providerOptions,
      schema: replicateImageModelOptionsSchema
    });
    let imageInputs = {};
    if (files != null && files.length > 0) {
      if (this.isFlux2Model) {
        for (let i = 0; i < Math.min(files.length, MAX_FLUX_2_INPUT_IMAGES); i++) {
          const key = i === 0 ? "input_image" : `input_image_${i + 1}`;
          imageInputs[key] = (0, import_provider_utils2.convertImageModelFileToDataUri)(files[i]);
        }
        if (files.length > MAX_FLUX_2_INPUT_IMAGES) {
          warnings.push({
            type: "other",
            message: `Flux-2 models support up to ${MAX_FLUX_2_INPUT_IMAGES} input images. Additional images are ignored.`
          });
        }
      } else {
        imageInputs = { image: (0, import_provider_utils2.convertImageModelFileToDataUri)(files[0]) };
        if (files.length > 1) {
          warnings.push({
            type: "other",
            message: "This Replicate model only supports a single input image. Additional images are ignored."
          });
        }
      }
    }
    let maskInput;
    if (mask != null) {
      if (this.isFlux2Model) {
        warnings.push({
          type: "other",
          message: "Flux-2 models do not support mask input. The mask will be ignored."
        });
      } else {
        maskInput = (0, import_provider_utils2.convertImageModelFileToDataUri)(mask);
      }
    }
    const { maxWaitTimeInSeconds, ...inputOptions } = replicateOptions != null ? replicateOptions : {};
    const preferHeader = maxWaitTimeInSeconds != null ? { prefer: `wait=${maxWaitTimeInSeconds}` } : { prefer: "wait" };
    const {
      value: { output },
      responseHeaders
    } = await (0, import_provider_utils2.postJsonToApi)({
      url: (
        // different endpoints for versioned vs unversioned models:
        version != null ? `${this.config.baseURL}/predictions` : `${this.config.baseURL}/models/${modelId}/predictions`
      ),
      headers: (0, import_provider_utils2.combineHeaders)(
        await (0, import_provider_utils2.resolve)(this.config.headers),
        headers,
        preferHeader
      ),
      body: {
        input: {
          prompt,
          aspect_ratio: aspectRatio,
          size,
          seed,
          num_outputs: n,
          ...imageInputs,
          ...maskInput != null ? { mask: maskInput } : {},
          ...inputOptions
        },
        // for versioned models, include the version in the body:
        ...version != null ? { version } : {}
      },
      successfulResponseHandler: (0, import_provider_utils2.createJsonResponseHandler)(
        replicateImageResponseSchema
      ),
      failedResponseHandler: replicateFailedResponseHandler,
      abortSignal,
      fetch: this.config.fetch
    });
    const outputArray = Array.isArray(output) ? output : [output];
    const images = await Promise.all(
      outputArray.map(async (url) => {
        const { value: image } = await (0, import_provider_utils2.getFromApi)({
          url,
          successfulResponseHandler: (0, import_provider_utils2.createBinaryResponseHandler)(),
          failedResponseHandler: replicateFailedResponseHandler,
          abortSignal,
          fetch: this.config.fetch
        });
        return image;
      })
    );
    return {
      images,
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders
      }
    };
  }
};
var replicateImageResponseSchema = import_v42.z.object({
  output: import_v42.z.union([import_v42.z.array(import_v42.z.string()), import_v42.z.string()])
});
var replicateImageModelOptionsSchema = (0, import_provider_utils2.lazySchema)(
  () => (0, import_provider_utils2.zodSchema)(
    import_v42.z.object({
      /**
       * Maximum time in seconds to wait for the prediction to complete in sync mode.
       * By default, Replicate uses sync mode with a 60-second timeout.
       *
       * - When not specified: Uses default 60-second sync wait (`prefer: wait`)
       * - When set to a positive number: Uses that duration (`prefer: wait=N`)
       */
      maxWaitTimeInSeconds: import_v42.z.number().positive().nullish(),
      /**
       * Guidance scale for classifier-free guidance.
       * Higher values make the output more closely match the prompt.
       */
      guidance_scale: import_v42.z.number().nullish(),
      /**
       * Number of denoising steps. More steps = higher quality but slower.
       */
      num_inference_steps: import_v42.z.number().nullish(),
      /**
       * Negative prompt to guide what to avoid in the generation.
       */
      negative_prompt: import_v42.z.string().nullish(),
      /**
       * Output image format.
       */
      output_format: import_v42.z.enum(["png", "jpg", "webp"]).nullish(),
      /**
       * Output image quality (1-100). Only applies to jpg and webp.
       */
      output_quality: import_v42.z.number().min(1).max(100).nullish(),
      /**
       * Strength of the transformation for img2img (0-1).
       * Lower values keep more of the original image.
       */
      strength: import_v42.z.number().min(0).max(1).nullish()
    }).passthrough()
  )
);

// src/replicate-video-model.ts
var import_provider = require("@ai-sdk/provider");
var import_provider_utils3 = require("@ai-sdk/provider-utils");
var import_v43 = require("zod/v4");
var ReplicateVideoModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v3";
    this.maxVideosPerCall = 1;
  }
  // Replicate video models support 1 video at a time
  get provider() {
    return this.config.provider;
  }
  async doGenerate(options) {
    var _a, _b, _c, _d, _e, _f, _g;
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const warnings = [];
    const replicateOptions = await (0, import_provider_utils3.parseProviderOptions)({
      provider: "replicate",
      providerOptions: options.providerOptions,
      schema: replicateVideoModelOptionsSchema
    });
    const [modelId, version] = this.modelId.split(":");
    const input = {};
    if (options.prompt != null) {
      input.prompt = options.prompt;
    }
    if (options.image != null) {
      if (options.image.type === "url") {
        input.image = options.image.url;
      } else {
        input.image = (0, import_provider_utils3.convertImageModelFileToDataUri)(options.image);
      }
    }
    if (options.aspectRatio) {
      input.aspect_ratio = options.aspectRatio;
    }
    if (options.resolution) {
      input.size = options.resolution;
    }
    if (options.duration) {
      input.duration = options.duration;
    }
    if (options.fps) {
      input.fps = options.fps;
    }
    if (options.seed) {
      input.seed = options.seed;
    }
    if (replicateOptions != null) {
      const opts = replicateOptions;
      if (opts.guidance_scale !== void 0 && opts.guidance_scale !== null) {
        input.guidance_scale = opts.guidance_scale;
      }
      if (opts.num_inference_steps !== void 0 && opts.num_inference_steps !== null) {
        input.num_inference_steps = opts.num_inference_steps;
      }
      if (opts.motion_bucket_id !== void 0 && opts.motion_bucket_id !== null) {
        input.motion_bucket_id = opts.motion_bucket_id;
      }
      if (opts.cond_aug !== void 0 && opts.cond_aug !== null) {
        input.cond_aug = opts.cond_aug;
      }
      if (opts.decoding_t !== void 0 && opts.decoding_t !== null) {
        input.decoding_t = opts.decoding_t;
      }
      if (opts.video_length !== void 0 && opts.video_length !== null) {
        input.video_length = opts.video_length;
      }
      if (opts.sizing_strategy !== void 0 && opts.sizing_strategy !== null) {
        input.sizing_strategy = opts.sizing_strategy;
      }
      if (opts.frames_per_second !== void 0 && opts.frames_per_second !== null) {
        input.frames_per_second = opts.frames_per_second;
      }
      if (opts.prompt_optimizer !== void 0 && opts.prompt_optimizer !== null) {
        input.prompt_optimizer = opts.prompt_optimizer;
      }
      for (const [key, value] of Object.entries(opts)) {
        if (![
          "pollIntervalMs",
          "pollTimeoutMs",
          "maxWaitTimeInSeconds",
          "guidance_scale",
          "num_inference_steps",
          "motion_bucket_id",
          "cond_aug",
          "decoding_t",
          "video_length",
          "sizing_strategy",
          "frames_per_second",
          "prompt_optimizer"
        ].includes(key)) {
          input[key] = value;
        }
      }
    }
    const maxWaitTimeInSeconds = replicateOptions == null ? void 0 : replicateOptions.maxWaitTimeInSeconds;
    const preferHeader = maxWaitTimeInSeconds != null ? { prefer: `wait=${maxWaitTimeInSeconds}` } : { prefer: "wait" };
    const predictionUrl = version != null ? `${this.config.baseURL}/predictions` : `${this.config.baseURL}/models/${modelId}/predictions`;
    const { value: prediction, responseHeaders } = await (0, import_provider_utils3.postJsonToApi)({
      url: predictionUrl,
      headers: (0, import_provider_utils3.combineHeaders)(
        await (0, import_provider_utils3.resolve)(this.config.headers),
        options.headers,
        preferHeader
      ),
      body: {
        input,
        ...version != null ? { version } : {}
      },
      successfulResponseHandler: (0, import_provider_utils3.createJsonResponseHandler)(
        replicatePredictionSchema
      ),
      failedResponseHandler: replicateFailedResponseHandler,
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    let finalPrediction = prediction;
    if (prediction.status === "starting" || prediction.status === "processing") {
      const pollIntervalMs = (_d = replicateOptions == null ? void 0 : replicateOptions.pollIntervalMs) != null ? _d : 2e3;
      const pollTimeoutMs = (_e = replicateOptions == null ? void 0 : replicateOptions.pollTimeoutMs) != null ? _e : 3e5;
      const startTime = Date.now();
      while (finalPrediction.status === "starting" || finalPrediction.status === "processing") {
        if (Date.now() - startTime > pollTimeoutMs) {
          throw new import_provider.AISDKError({
            name: "REPLICATE_VIDEO_GENERATION_TIMEOUT",
            message: `Video generation timed out after ${pollTimeoutMs}ms`
          });
        }
        await (0, import_provider_utils3.delay)(pollIntervalMs);
        if ((_f = options.abortSignal) == null ? void 0 : _f.aborted) {
          throw new import_provider.AISDKError({
            name: "REPLICATE_VIDEO_GENERATION_ABORTED",
            message: "Video generation request was aborted"
          });
        }
        const { value: statusPrediction } = await (0, import_provider_utils3.getFromApi)({
          url: finalPrediction.urls.get,
          headers: await (0, import_provider_utils3.resolve)(this.config.headers),
          successfulResponseHandler: (0, import_provider_utils3.createJsonResponseHandler)(
            replicatePredictionSchema
          ),
          failedResponseHandler: replicateFailedResponseHandler,
          abortSignal: options.abortSignal,
          fetch: this.config.fetch
        });
        finalPrediction = statusPrediction;
      }
    }
    if (finalPrediction.status === "failed") {
      throw new import_provider.AISDKError({
        name: "REPLICATE_VIDEO_GENERATION_FAILED",
        message: `Video generation failed: ${(_g = finalPrediction.error) != null ? _g : "Unknown error"}`
      });
    }
    if (finalPrediction.status === "canceled") {
      throw new import_provider.AISDKError({
        name: "REPLICATE_VIDEO_GENERATION_CANCELED",
        message: "Video generation was canceled"
      });
    }
    const videoUrl = finalPrediction.output;
    if (!videoUrl) {
      throw new import_provider.AISDKError({
        name: "REPLICATE_VIDEO_GENERATION_ERROR",
        message: "No video URL in response"
      });
    }
    return {
      videos: [
        {
          type: "url",
          url: videoUrl,
          mediaType: "video/mp4"
        }
      ],
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders
      },
      providerMetadata: {
        replicate: {
          videos: [
            {
              url: videoUrl
            }
          ],
          predictionId: finalPrediction.id,
          metrics: finalPrediction.metrics
        }
      }
    };
  }
};
var replicatePredictionSchema = import_v43.z.object({
  id: import_v43.z.string(),
  status: import_v43.z.enum(["starting", "processing", "succeeded", "failed", "canceled"]),
  output: import_v43.z.string().nullish(),
  error: import_v43.z.string().nullish(),
  urls: import_v43.z.object({
    get: import_v43.z.string()
  }),
  metrics: import_v43.z.object({
    predict_time: import_v43.z.number().nullish()
  }).nullish()
});
var replicateVideoModelOptionsSchema = (0, import_provider_utils3.lazySchema)(
  () => (0, import_provider_utils3.zodSchema)(
    import_v43.z.object({
      pollIntervalMs: import_v43.z.number().positive().nullish(),
      pollTimeoutMs: import_v43.z.number().positive().nullish(),
      maxWaitTimeInSeconds: import_v43.z.number().positive().nullish(),
      guidance_scale: import_v43.z.number().nullish(),
      num_inference_steps: import_v43.z.number().nullish(),
      motion_bucket_id: import_v43.z.number().nullish(),
      cond_aug: import_v43.z.number().nullish(),
      decoding_t: import_v43.z.number().nullish(),
      video_length: import_v43.z.string().nullish(),
      sizing_strategy: import_v43.z.string().nullish(),
      frames_per_second: import_v43.z.number().nullish(),
      prompt_optimizer: import_v43.z.boolean().nullish()
    }).loose()
  )
);

// src/version.ts
var VERSION = true ? "2.0.33" : "0.0.0-test";

// src/replicate-provider.ts
function createReplicate(options = {}) {
  const getHeaders = () => (0, import_provider_utils4.withUserAgentSuffix)(
    {
      Authorization: `Bearer ${(0, import_provider_utils4.loadApiKey)({
        apiKey: options.apiToken,
        environmentVariableName: "REPLICATE_API_TOKEN",
        description: "Replicate"
      })}`,
      ...options.headers
    },
    `ai-sdk/replicate/${VERSION}`
  );
  const createImageModel = (modelId) => {
    var _a;
    return new ReplicateImageModel(modelId, {
      provider: "replicate",
      baseURL: (_a = options.baseURL) != null ? _a : "https://api.replicate.com/v1",
      headers: getHeaders(),
      fetch: options.fetch
    });
  };
  const createVideoModel = (modelId) => {
    var _a;
    return new ReplicateVideoModel(modelId, {
      provider: "replicate.video",
      baseURL: (_a = options.baseURL) != null ? _a : "https://api.replicate.com/v1",
      headers: getHeaders,
      fetch: options.fetch
    });
  };
  const embeddingModel = (modelId) => {
    throw new import_provider2.NoSuchModelError({
      modelId,
      modelType: "embeddingModel"
    });
  };
  return {
    specificationVersion: "v3",
    image: createImageModel,
    imageModel: createImageModel,
    languageModel: (modelId) => {
      throw new import_provider2.NoSuchModelError({
        modelId,
        modelType: "languageModel"
      });
    },
    embeddingModel,
    textEmbeddingModel: embeddingModel,
    video: createVideoModel,
    videoModel: createVideoModel
  };
}
var replicate = createReplicate();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VERSION,
  createReplicate,
  replicate
});
